java/src/main/kotlin/io/spine/validation/java/PatternFieldGenerator.kt

```diff
@@ -0,0 +1,226 @@
+import io.spine.protodata.java.Expression
+import io.spine.protodata.java.FieldDeclaration
+import io.spine.protodata.java.InitField
+import io.spine.protodata.java.Literal
+import io.spine.protodata.java.MethodCall
+import io.spine.protodata.java.MethodDeclaration
+import io.spine.protodata.java.ReadVar
+import io.spine.protodata.java.StringLiteral
+import io.spine.protodata.java.call
+import io.spine.protodata.java.field
+import io.spine.protodata.java.toBuilder
+import io.spine.validate.ConstraintViolation
+import io.spine.validation.PATTERN
+import io.spine.validation.PatternField
+import io.spine.validation.isRepeatedString
+import io.spine.validation.isSingularString
+import io.spine.validation.java.ErrorPlaceholder.FIELD_PATH
+import io.spine.validation.java.ErrorPlaceholder.FIELD_TYPE
+import io.spine.validation.java.ErrorPlaceholder.FIELD_VALUE
+import io.spine.validation.java.ErrorPlaceholder.PARENT_TYPE
+import io.spine.validation.java.ErrorPlaceholder.REGEX_MODIFIERS
+import io.spine.validation.java.ErrorPlaceholder.REGEX_PATTERN
+import io.spine.validation.java.MessageValidationRenderer.MessageScope.message
+import io.spine.validation.java.MessageValidationRenderer.ValidateScope.parentPath
+import io.spine.validation.java.MessageValidationRenderer.ValidateScope.violations
+import java.util.regex.Matcher
+import java.util.regex.Pattern
+import org.apache.commons.lang.StringEscapeUtils.escapeJava
+
+/**
+ * The generator for `(pattern)` option.
+ *
+ * Generates code for a single field represented by the provided [view].
+ */
+internal class PatternFieldGenerator(private val view: PatternField) {
+
+    private val field = view.subject
+    private val fieldType = field.type
+    private val fieldAccess = message.field(field)
+    private val declaringType = field.declaringType
+    private val camelFieldName = field.name.camelCase
+    private val pattern = compilePattern()
+
+    /**
+     * Generates code for a field represented by the [view].
+     */
+    fun generate(): FieldOptionCode = when {
+        fieldType.isSingularString -> {
+            val fieldValue = fieldAccess.getter<String>()
+            val constraint = singularStringConstraint(fieldValue)
+            FieldOptionCode(constraint, pattern.field)
+        }
+
+        fieldType.isRepeatedString -> {
+            val fieldValue = fieldAccess.getter<List<String>>()
+            val validateRepeatedField = mangled("validate$camelFieldName")
+            val validateRepeatedFieldDecl = validateRepeatedField(fieldValue, validateRepeatedField)
+            val constraint = repeatedStringConstraint(fieldValue, validateRepeatedField)
+            FieldOptionCode(constraint, pattern.field, validateRepeatedFieldDecl)
+        }
+
+        else -> error {
+            "Unsupported field type: `${fieldType.name}`. The `(${PATTERN})` option can be" +
+                    "applied only to singular or repeated string fields."
+        }
```

Comment:
I have tried to get rid of this nonsense check because the policy already confirmed the field type. Couldn't find a solution. Information about the confirmed field type can arrive in the event, but Proto enums inevitably have two `service` constants, leading to the same `else` branch in exhaustive `when`.